Single Sign On with Github

shivam-Purohit opened the pull request:
closes #1718

# Description
Single-sign-on Github feature for the app

## Type of change

Please delete options that are not relevant.


- [x] New feature (non-breaking change which adds functionality).


# Checklist

- [x] I have read and understood the rules about [how to Contribute](https://intelowl.readthedocs.io/en/latest/Contribute.html) to this project
- [x] The pull request is for the branch `develop`
- [ ] A new plugin (analyzer, connector, visualizer or playbook) was added or changed, in which case:
    - [ ] [Usage](https://github.com/intelowlproject/IntelOwl/blob/master/docs/source/Usage.md) file was updated.
    - [ ] [Advanced-Usage](./Advanced-Usage.md) was updated (in case the plugin provides additional optional configuration).
    - [ ] If the plugin requires mocked testing, `_monkeypatch()` was used in its class to apply the necessary decorators.
    - [ ] If a File analyzer was added and it supports a mimetype which is not already supported, you added a sample of that type inside the archive `test_files.zip` and you added the default tests for that mimetype in [test_classes.py](https://github.com/intelowlproject/IntelOwl/blob/master/tests/analyzers_manager/test_classes.py).
    - [ ] If you created a new analyzer and it is free (does not require API keys), please add it in the `FREE_TO_USE_ANALYZERS` playbook in `playbook_config.json`.
    - [ ] Check if it could make sense to add that analyzer/connector to other freely available playbooks.
    - [ ] I have provided the resulting raw JSON of a finished analysis and a screenshot of the results.
- [ ] If external libraries/packages with restrictive licenses were used, they were added in the [Legal Notice](https://github.com/certego/IntelOwl/blob/master/.github/legal_notice.md) section.
- [x] Linters (`Black`, `Flake`, `Isort`) gave 0 errors. If you have correctly installed [pre-commit](https://intelowl.readthedocs.io/en/latest/Contribute.html#how-to-start-setup-project-and-development-instance), it does these checks and adjustments on your behalf.
- [ ] I have added tests for the feature/bug I solved (see `tests` folder). All the tests (new and old ones) gave 0 errors.
- [ ] If changes were made to an existing model/serializer/view, the docs were updated and regenerated (check [CONTRIBUTE.md](./Contribute.md)).
- [x] If the GUI has been modified:
    - [x] I have a provided a screenshot of the result in the PR.
    - [ ] I have created new frontend tests for the new component or updated existing ones.

### Important Rules
![ubuntu project  Running  - Oracle VM VirtualBox 29-06-2023 14_25_27](https://github.com/intelowlproject/IntelOwl/assets/91889807/76bc290d-e0b5-4313-bbc8-56022b0ca8f8)

mlodic commented on intel_owl/settings/__init__.py:
```diff
@@ -16,6 +16,8 @@
     "django.contrib.messages",
     "django.contrib.staticfiles",
     "django.contrib.postgres",
+    # sites
```
could you specify here as a comment that this package is required for the `allauth` installation? otherwise people in future could wonder why it is here.

mlodic commented on frontend/src/components/auth/Login.jsx:
```diff
@@ -90,6 +92,56 @@ export default function Login() {
+                }}
+              >
+                <img
+                  src={`${PUBLIC_URL}/icons8-github-48.png`}
```
in the dark mode of IntelOwl, I think it is better to use the icon with opposite colors (white as main background)

mlodic commented on intel_owl/settings/auth.py:
```diff
@@ -54,3 +54,15 @@
+SOCIALACCOUNT_PROVIDERS = {
+    "github": {
+        "APP": {
+            "client_id": "4a7aed5c9ee5a88afb5b",
+            "secret": "a486ba124a601920b559c472631d90a84863f83c",
```
If these are the credentials that you are using right now in your test app, I suggest to remove them and rotate them.
These secrets should always be extracted from environment variables, and never hardcoded here, even for testing purposes only. Otherwise it is too easy to risk to push them.

Plus, we would need to add a part in the documentation where we explain how the IntelOwl admin would need to setup these secrets to make them work in a production environment.
And also few lines about how you can test the integration locally. ( this would also help me to do a test as a user and validate the work you have done)

mlodic commented on frontend/src/components/auth/Login.jsx:
```diff
@@ -90,6 +92,56 @@ export default function Login() {
             <h3 className="fw-bold col-auto me-auto mt-2">Log In</h3>
+            <div className="col-auto">
+              <a
+                href={`${AUTH_BASE_URI}/gihutb`}
```
typo?

mlodic commented on authentication/views.py:
```diff
@@ -260,3 +266,42 @@ def check_registration_setup(request):
+                request, user, "allauth.account.auth_backends.AuthenticationBackend"
+            )
+
+            return redirect(self.request.build_absolute_uri(f"/login?token={token}"))
```
I don't think the token should be bassed as a parameter, more likely as a header

mlodic reviewed (commented):
just made few comments after having read this.

I leave a reference to Github where they explain how the Oauth2 works: https://docs.github.com/en/apps/oauth-apps/building-oauth-apps/authorizing-oauth-apps

shivam-Purohit commented:
@mlodic I was able to exchange the code with the accesscode or accesstoken for the github. I just printed that on the screen for now. From here the token should be used to login into the app, right? I will try to look into google sso how we did that

mlodic commented:
yep, now you would need to use that accesstoken to login to the application. You would need to do something similar at what is implemented in `validate_and_return_user` for google

shivam-Purohit commented:
> yep, now you would need to use that accesstoken to login to the application. You would need to do something similar at what is implemented in `validate_and_return_user` for google

when i try to implement something similar as google it throws an error

`detail | 'Method "GET" not allowed.'`

This is the code i am using 
```
class GitHubLoginCallbackView(LoginView):
    def get_accesstoken(request):
        code = request.GET.get("code")

        url = "https://github.com/login/oauth/access_token"

        params = {
            "client_id": secrets.get_secret("GITHUB_CLIENT_ID"),
            "client_secret": secrets.get_secret("GITHUB_CLIENT_SECRET"),
            "code": code,
            "redirect_uri": "http://localhost/accounts/github/login/callback/",
        }

        response = requests.post(url, params=params)
        if response.status_code == 200:
            return response.text

        return HttpResponse("failed")
```

i found error similar to this in this [issue](https://github.com/Tivix/django-rest-auth/issues/357)
I will try to come up with something different like creating two different functions, one for getting the accesstoken and other to login to the app using the accesstoken

shivam-Purohit commented:
This is the error i am currently facing
```
duplicate key value violates unique constraint "certego_saas_user_user_username_key"
DETAIL:  Key (username)=(Shivam Purohit) already exists.
```

I tried using different account but same error occurs

mlodic commented on authentication/views.py:
```diff
@@ -267,20 +269,80 @@ def check_registration_setup(request):
+class GitHubLoginCallbackView(LoginView):
 
-    url = "https://github.com/login/oauth/access_token"
+    access_token = None
```
why do you need this class attribute?

mlodic commented on authentication/views.py:
```diff
@@ -267,20 +269,80 @@ def check_registration_setup(request):
+
+    @staticmethod
+    def validate_and_return_user(request, access_token):
+        try:
+            token = access_token
+        except (
+            OAuthError,
+            OAuth2Error,
+        ):
+            # Not giving out the actual error as we risk exposing the client secret
+            raise AuthenticationFailed("OAuth authentication error.")
```
you know that this does not make any sense? you can't just copy/paste randomly :P

mlodic commented on authentication/views.py:
```diff
@@ -267,20 +269,80 @@ def check_registration_setup(request):
+            "Accept": "application/vnd.github.v3+json",
+        }
+        response = requests.get("https://api.github.com/user", headers=headers)
+        user_data = response.json()
```
can you write down what do you get from the APIs at this time? you need to add more logs to understand what is happening.

mlodic reviewed (commented):
please share complete stack traces and not only the message log otherwise it is harder for us to understand where the error happened

mlodic commented:
also, please resolve the conflicts in this branch by pulling changes from the `develop` branch here. This is important because otherwise you risk to write code which creates additional conflicts.

shivam-Purohit commented:
> please share complete stack traces and not only the message log otherwise it is harder for us to understand where the error happened

```
2023-07-03 14:44:13.513 UTC [62] ERROR:  duplicate key value violates unique constraint "certego_saas_user_user_username_key"

intelowl_postgres               | 2023-07-03 14:44:13.513 UTC [62] DETAIL:  Key (username)=(Shivam Purohit) already exists.

intelowl_postgres               | 2023-07-03 14:44:13.513 UTC [62] STATEMENT:  INSERT INTO "certego_saas_user_user" ("password", "last_login", "is_superuser", "username", "first_name", "last_name", "email", "is_staff", "is_active", "date_joined", "approved") VALUES ('!Clo51SKKDilc6wGDTYPlen8l5LKPaqTNr1ZX0Myd', NULL, false, 'Shivam Purohit', '', '', '', false, true, '2023-07-03T14:44:13.508375+00:00'::timestamptz, NULL) RETURNING "certego_saas_user_user"."id"

intelowl_uwsgi                  | ERROR:django.request:Internal Server Error: /accounts/github/login/callback/

intelowl_uwsgi                  | Traceback (most recent call last):

intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 318, in validate_and_return_user

intelowl_uwsgi                  |     return User.objects.get(email=user_email)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method

intelowl_uwsgi                  |     return getattr(self.get_queryset(), name)(*args, **kwargs)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 650, in get

intelowl_uwsgi                  |     raise self.model.DoesNotExist(

intelowl_uwsgi                  | certego_saas.apps.user.models.User.DoesNotExist: User matching query does not exist.

intelowl_uwsgi                  | 

intelowl_uwsgi                  | During handling of the above exception, another exception occurred:

intelowl_uwsgi                  | 

intelowl_uwsgi                  | Traceback (most recent call last):

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute

intelowl_uwsgi                  |     return self.cursor.execute(sql, params)

intelowl_uwsgi                  | psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "certego_saas_user_user_username_key"

intelowl_uwsgi                  | DETAIL:  Key (username)=(Shivam Purohit) already exists.

intelowl_uwsgi                  | 

intelowl_uwsgi                  | 

intelowl_uwsgi                  | The above exception was the direct cause of the following exception:

intelowl_uwsgi                  | 

intelowl_uwsgi                  | Traceback (most recent call last):

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/exception.py", line 56, in inner

intelowl_uwsgi                  |     response = get_response(request)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/core/handlers/base.py", line 197, in _get_response

intelowl_uwsgi                  |     response = wrapped_callback(request, *callback_args, **callback_kwargs)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/views/decorators/csrf.py", line 55, in wrapped_view

intelowl_uwsgi                  |     return view_func(*args, **kwargs)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/views/generic/base.py", line 103, in view

intelowl_uwsgi                  |     return self.dispatch(request, *args, **kwargs)

intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 299, in dispatch

intelowl_uwsgi                  |     return self.validate_and_return_user(request, access_token)

intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 321, in validate_and_return_user

intelowl_uwsgi                  |     return User.objects.create_user(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/models.py", line 161, in create_user

intelowl_uwsgi                  |     return self._create_user(username, email, password, **extra_fields)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/models.py", line 155, in _create_user

intelowl_uwsgi                  |     user.save(using=self._db)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/base_user.py", line 69, in save

intelowl_uwsgi                  |     super().save(*args, **kwargs)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 812, in save

intelowl_uwsgi                  |     self.save_base(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 863, in save_base

intelowl_uwsgi                  |     updated = self._save_table(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1006, in _save_table

intelowl_uwsgi                  |     results = self._do_insert(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/base.py", line 1047, in _do_insert

intelowl_uwsgi                  |     return manager._insert(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method

intelowl_uwsgi                  |     return getattr(self.get_queryset(), name)(*args, **kwargs)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 1791, in _insert

intelowl_uwsgi                  |     return query.get_compiler(using=using).execute_sql(returning_fields)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1660, in execute_sql

intelowl_uwsgi                  |     cursor.execute(sql, params)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 102, in execute

intelowl_uwsgi                  |     return super().execute(sql, params)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 67, in execute

intelowl_uwsgi                  |     return self._execute_with_wrappers(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers

intelowl_uwsgi                  |     return executor(sql, params, many, context)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute

intelowl_uwsgi                  |     return self.cursor.execute(sql, params)

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/utils.py", line 91, in __exit__

intelowl_uwsgi                  |     raise dj_exc_value.with_traceback(traceback) from exc_value

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/db/backends/utils.py", line 89, in _execute

intelowl_uwsgi                  |     return self.cursor.execute(sql, params)

intelowl_uwsgi                  | django.db.utils.IntegrityError: duplicate key value violates unique constraint "certego_saas_user_user_username_key"

intelowl_uwsgi                  | DETAIL:  Key (username)=(Shivam Purohit) already exists.
```

mlodic commented on authentication/views.py:
```diff
@@ -302,7 +299,8 @@ def dispatch(self, request, *args, **kwargs):
             # Handle case when code is not present
-            return HttpResponse("Code not found")
+            logger.warning("Failed to retrive code")
+            return HttpResponse("OAuth authentication failed", status=400)
```
you should use constant from `http.status` library instead of raw numbers

mlodic commented on authentication/views.py:
```diff
@@ -306,27 +306,27 @@ def dispatch(self, request, *args, **kwargs):
     def validate_and_return_user(request, access_token):
         try:
             token = access_token
+            headers = {
+                "Authorization": f"Bearer {token}",
+                "Accept": "application/vnd.github.v3+json",
+            }
+            response = requests.get("https://api.github.com/user", headers=headers)
+            user_data = response.json()
+            user_name = user_data.get("name")
+            user_email = user_data.get("email")
+            try:
+                return User.objects.get(email=user_email)
+            except User.DoesNotExist:
+                logging.info("[Github Oauth] User does not exist. Creating new user.")
+                return User.objects.create_user(
+                    email=user_email, username=user_name, password=None
+                )
```
what is this change? what do you wanted to do here? cause I don't see the logic

mlodic commented on authentication/views.py:
```diff
@@ -306,27 +306,27 @@ def dispatch(self, request, *args, **kwargs):
     def validate_and_return_user(request, access_token):
         try:
             token = access_token
```
in which way this is useful?

mlodic commented on authentication/views.py:
```diff
@@ -306,27 +306,27 @@ def dispatch(self, request, *args, **kwargs):
+                )
         except (
             OAuthError,
             OAuth2Error,
```
again, we can't just copy/paste and leave the code like that is. Why do we need to keep using this exceptions handling for this case?

mlodic commented:
```
intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 299, in dispatch

intelowl_uwsgi                  |     return self.validate_and_return_user(request, access_token)

intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 321, in validate_and_return_user

intelowl_uwsgi                  |     return User.objects.create_user(

intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/models.py", line 161, in create_user

intelowl_uwsgi                  |     return self._create_user(username, email, password, **extra_fields)
```
Well, it seems that you already have a user registered with that username...have you tried different emails? Have you checked the database to understand whether there is another "Shivam" and how did it happen?

shivam-Purohit commented:
> ```
> intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 299, in dispatch
> 
> intelowl_uwsgi                  |     return self.validate_and_return_user(request, access_token)
> 
> intelowl_uwsgi                  |   File "/opt/deploy/intel_owl/authentication/views.py", line 321, in validate_and_return_user
> 
> intelowl_uwsgi                  |     return User.objects.create_user(
> 
> intelowl_uwsgi                  |   File "/usr/local/lib/python3.9/site-packages/django/contrib/auth/models.py", line 161, in create_user
> 
> intelowl_uwsgi                  |     return self._create_user(username, email, password, **extra_fields)
> ```
> 
> Well, it seems that you already have a user registered with that username...have you tried different emails? Have you checked the database to understand whether there is another "Shivam" and how did it happen?

I tried using a different github account that wasn't present in the database. But still, it gives the same error. Which should not have been the case. But it should capture the user if the user already exists in the account.

mlodic commented:
As I have already told you, you need to debug the problem by adding other lines of logs and by checking the database entries

shivam-Purohit commented:
> As I have already told you, you need to debug the problem by adding other lines of logs and by checking the database entries

The problem is in receiving from the user_name and user_email from the request. Its returning None.

shivam-Purohit commented:
I will be trying to setup the `github_login `view by myself. I think the error has something to do with the `callback_uri `which is really confusing me

mlodic commented:
ok but could you please log the data that you gets from the APIs as I asked you?
```
            user_data = response.json()
            user_name = user_data.get("name")
            user_email = user_data.get("email")
```
The email is not mandatory as you can see from the Django docs: https://docs.djangoproject.com/en/4.2/ref/contrib/auth/#django.contrib.auth.models.UserManager.create_user

shivam-Purohit commented:
> ok but could you please log the data that you gets from the APIs as I asked you?
> 
> ```
>             user_data = response.json()
>             user_name = user_data.get("name")
>             user_email = user_data.get("email")
> ```
> 
> The email is not mandatory as you can see from the Django docs: https://docs.djangoproject.com/en/4.2/ref/contrib/auth/#django.contrib.auth.models.UserManager.create_user

so that shouldn't be the problem then. Would it be okay if I push those in about 2 hrs?

shivam-Purohit commented:
The feature is working now. I removed the logic from the frontend button. I will replace with some other that suits github request.
maybe a simpler one that just checks if the response is ok. I will also change the github logo, will try to find something with white background.

shivam-Purohit commented:
> also, please resolve the conflicts in this branch by pulling changes from the `develop` branch here. This is important because otherwise you risk to write code which creates additional conflicts.

I merged the `develop` into `sso-github ` . It has bunch of changes from develop now and some warning. The original feature is working fine and the app is not breaking , i am not sure about the other functionalities . What should i do should i push the changes? or should i resolve in the github UI?

mlodic commented:
> I merged the develop into sso-github  . It has bunch of changes from develop now and some warning. The original feature is working fine and the app is not breaking , i am not sure about the other functionalities . What should i do should i push the changes? or should i resolve in the github UI?

I see you already did it but yeah you just need to review the changes made from the upstream and see whether there is something or not that collides with what you have written. In that case, if that's easy, you can try to solve on your own. If that's not, you should ask for help to the others that wrote the feature. Then, you are sure that old functionalities works only if the previous writer has written the correct tests and this is why tests are important

mlodic commented on authentication/views.py:
```diff
@@ -275,3 +281,63 @@ def checkConfiguration(request):
+                    return User.objects.create_user(
+                        email=user_email, username=user_name, password=None
+                    )
+        raise AuthenticationFailed("GitHub OAuth authentication error.")
```
good usage of raise and the exception

mlodic commented on authentication/views.py:
```diff
@@ -275,3 +281,63 @@ def checkConfiguration(request):
+                user_data = response.json()
+                user_name = user_data.get("login")
+                user_email = user_data.get("email")
+                logging.info(f"received userinfo from github: {response}")
```
have you at least read which is the output of this log? cause in the case you really had, you would have understood that this cannot be the proper variable to use.

Plus, are we sure that we can log this user info? cause there is sensitive information. I would just use a `debug` log for debugging purposes because that is the reason why this log can become useful to us. To understand what is happening while we are developing

mlodic reviewed (commented):
> The feature is working now. I removed the logic from the frontend button. I will replace with some other that suits github request.
maybe a simpler one that just checks if the response is ok. I will also change the github logo, will try to find something with white background.

great! once you have something that could be tested please ping that I'll test it

shivam-Purohit commented:
> I see you already did it but yeah you just need to review the changes made from the upstream and see whether there is something or not that collides with what you have written. In that case, if that's easy, you can try to solve on your own. If that's not, you should ask for help to the others that wrote the feature. Then, you are sure that old functionalities works only if the previous writer has written the correct tests and this is why tests are important

there was some conflicts on the new RE_CAPTCHA commits. I merged the newer ones. Though the errors that i am receiving in the terminal are from the playbook manager and other apps.
something like this
```
ntelowl_postgres               | 2023-07-07 13:19:37.044 UTC [205] ERROR:  column visualizers_manager_visualizerconfig.playbook_id does not exist at character 273

intelowl_postgres               | 2023-07-07 13:19:37.044 UTC [205] STATEMENT:  EXPLAIN SELECT "visualizers_manager_visualizerconfig"."name", "visualizers_manager_visualizerconfig"."description", "visualizers_manager_visualizerconfig"."disabled", "visualizers_manager_visualizerconfig"."python_module", "visualizers_manager_visualizerconfig"."config", "visualizers_manager_visualizerconfig"."playbook_id" FROM "visualizers_manager_visualizerconfig" WHERE "visualizers_manager_visualizerconfig"."playbook_id" = 'FREE_TO_USE_ANALYZERS' ORDER BY "visualizers_manager_visualizerconfig"."name" ASC, "visualizers_manager_visualizerconfig"."disabled" ASC LIMIT 21


```

if there is something wrong i can revert back to previous commit.

shivam-Purohit commented:
Also what should i add in the frontend to check if the oauth is configured or not like the google?should i check if the oauth variables exists in the setting or env ?

mlodic commented:
> ntelowl_postgres               | 2023-07-07 13:19:37.044 UTC [205] ERROR:  column visualizers_manager_visualizerconfig.playbook_id does not exist at character 273

If that happens it is because some migrations were probably not applied correctly. Can you try to check the output of `showmigrations` to understand which migrations was not applied and check if there are any errors when running IntelOwl? 
In case the error persist, the easies thing is to drop all the database by using `down -v` and start from scratch

> should i check if the oauth variables exists in the setting or env ?
yep, we need an endpoint in IntelOwl which checks the configuration. It should behave like Google auth already behaves

shivam-Purohit commented:
> If that happens it is because some migrations were probably not applied correctly. Can you try to check the output of `showmigrations` to understand which migrations was not applied and check if there are any errors when running IntelOwl?
> In case the error persist, the easies thing is to drop all the database by using `down -v` and start from scratch

Yeah few migrations were not applied, ig the error is not there now. But a bunch of log files are shown at the screen
`intelowl_uwsgi                  | DEBUG:api_app.core.serializers:Parameter timeout for plugin XForceExchange has value 5 for user shivam-Purohit`

should this be happening?

> yep, we need an endpoint in IntelOwl which checks the configuration. It should behave like Google auth already behaves

okay i will try to check if the variables are set. if not i will try to return an error to the frontend 
if not we can redirect to the page

shivam-Purohit commented:
I added the check. One thing is that unlike google the response is not returning any error even if the client_id is not set, It redirects to the github page with the empty string as client_id. And returns 404
Should I remove that check, I added a separate one to check for client_id.


I think you can try the feature now. I will try to fix these CI issues

shivam-Purohit commented:
Tested the app locally, and it did showed some errors but not related to my change. I checked the migrations and that's okay.

mlodic commented:
I am testing it now manually, but, you know, we need unit tests too otherwise we can't integrate this feature

Please check `test_oauth` where one of the previous GSoCers implemented the tests for the Google Oauth.

mlodic reviewed (changes requested):
The authentication works!

however there are no tests and no documentation so this makes difficult even to test this manually and understand that things work. People would need to set up a Github app and there is no help in that sense. It is very important to specify the callback required to have things work properly.

shivam-Purohit commented:
> The authentication works!
> 
> however there are no tests and no documentation so this makes difficult even to test this manually and understand that things work. People would need to set up a Github app and there is no help in that sense. It is very important to specify the callback required to have things work properly.

I will add the documentation and test as you mentioned. In the documentation I will keep these points
1. the link to github new app
2. what should be the callback value
3. setup the client_id and client_secret

shivam-Purohit commented:
i tried the test manually. Works to most extent.  we can also test for a random string that we can send in our initial request to the github page and get it back from the final response with access_token. Should we add that?

mlodic commented:
Tests you pushed are failing if you check the CI, they are not working

```
======================================================================
FAIL: test_github_enabled (tests.auth.test_oauth.TestGitHubOAuth)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/deploy/intel_owl/tests/auth/test_oauth.py", line 88, in test_github_enabled
    self.assertEqual(response.status_code, 302)
AssertionError: 401 != 302

----------------------------------------------------------------------
Ran 253 tests in 375.927s
```

mlodic commented:
> I maybe wrong but shouldn't they be failing as it is not configured? When I tested the tests, the github_enable was failing when it was not configured and github_disable is failing when it is configured.
If so should those tests be skipped in CI?

no, the tests are meant to be always successful. You should use a simple scientific approach.
What do you want to prove? Create your hypothesis and test it if that works.
In this case, if you want to test when the Github oauth is enabled, you should first change the test environment to have the setting enabled like it should and then test if that works with the right environment. You can and you should manipolate the environment, the db, anything during tests to test what you want to prove.

shivam-Purohit commented:
> no, the tests are meant to be always successful. You should use a simple scientific approach.
> What do you want to prove? Create your hypothesis and test it if that works.
> In this case, if you want to test when the Github oauth is enabled, you should first change the test environment to have the setting enabled like it should and then test if that works with the right environment. You can and you should manipolate the environment, the db, anything during tests to test what you want to prove.

yes, talking for the ci i add the env variables for it. But i don't have the access to set the actual values. If they arent set the test shall fail right?

In the local if I don't configure the google oauth, its also failing indicating that the oauth is not configured.
FAIL: test_github_disabled (tests.auth.test_oauth.TestGitHubOAuth)
```

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/opt/deploy/intel_owl/tests/auth/test_oauth.py", line 83, in test_github_disabled

    self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)

AssertionError: 302 != 401



======================================================================

FAIL: test_google_enabled (tests.auth.test_oauth.TestOAuth)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/opt/deploy/intel_owl/tests/auth/test_oauth.py", line 41, in test_google_enabled

    self.assertEqual(response.status_code, 302)

AssertionError: 401 != 302


```

this is what I wanted to convey

shivam-Purohit commented:
I took care of the toolkit bug. It was using the same variables to trigger the action so it was coinciding. I separated the STATE variables so it should work fine now.

also how do i pass the deepsource python check and gitguardian check
gitguardian can be fixed if we squash the commit at the end i guess

mlodic commented:
> yes, talking for the ci i add the env variables for it. But i don't have the access to set the actual values. If they arent set the test shall fail right?
In the local if I don't configure the google oauth, its also failing indicating that the oauth is not configured.
FAIL: test_github_disabled (tests.auth.test_oauth.TestGitHubOAuth)

again. Tests are not meant to fail anywhere. Or, if you want the test to fail you should check it with assertFail()

> gitguardian can be fixed if we squash the commit at the end i guess
yes

then deepsource is detecting an issue in identation, it could be your change in that file to cause it

mlodic commented on requirements/project-requirements.txt:
```diff
@@ -54,7 +54,7 @@ pyelftools==0.29
 pyhashlookup==1.2.0
 pyimpfuzzy==0.5
-pymisp==2.4.168.1
+pymisp>=2.4.168.1,<2.5.0
```
that is wrong, please revert this. Anywhere we use `<` and `>` for dependencies, why we should add that here?

It is important to pick a single version otherwise we risk unattended consequences in deployments.

we could pin the last version 2.4.172 and see if that dependency errors is solved.

shivam-Purohit commented:
> no, the tests are meant to be always successful. You should use a simple scientific approach.
> What do you want to prove? Create your hypothesis and test it if that works.
> In this case, if you want to test when the Github oauth is enabled, you should first change the test environment to have the setting enabled like it should and then test if that works with the right environment. You can and you should manipolate the environment, the db, anything during tests to test what you want to prove.

Then If a want to check for github_disable I should first check if the configuration exists or not before moving to the other checks right? This way I will be able to check for both at the same time.

shivam-Purohit commented:
so I had to destroy the client_id so that it could check for the github_disable. I think it should work fine now.
Also, that deepsource check, even if I change the code now, it wont push as nothing is changed eventually. I will look into it, what can i do

mlodic commented on docs/source/Advanced-Configuration.md:
```diff
@@ -67,6 +68,12 @@ While configuring Google Auth2 you can choose either to enable access to the all
 
+#### GitHub OAuth
+
+Create a new OAuth app on [GitHub](https://github.com/settings/applications/new), set the `Authorization callback URL` as the `http://localhost/api/auth/github-callback` and then generate the `Client ID` and `Client secret`.
```
That is not entirely correct, because localhost would work only if you are running the application locally. We should say that `http://<your_domain|localhost>/api/auth/github-callback` instead

mlodic commented on requirements/project-requirements.txt:
```diff
@@ -54,7 +54,7 @@ pyelftools==0.29
 pyhashlookup==1.2.0
 pyimpfuzzy==0.5
-pymisp==2.4.168.1
+pymisp>=2.4.172
```
I meant to set == like everywhere else

mlodic commented on tests/auth/test_oauth.py:
```diff
@@ -72,3 +72,49 @@ def test_google_callback(self, mock_validate_and_return_user: Mock):
+    github_auth_uri = reverse("oauth_github")
+    github_auth_callback_uri = reverse("oauth_github_callback")
+
+    @patch("authentication.views.secrets.get_secret")
+    def test_github_disabled(self, mock_get_secret):
+        mock_get_secret.return_value = None
```
you should patch only github variables, not every secret

mlodic commented on tests/auth/test_oauth.py:
```diff
@@ -72,3 +72,49 @@ def test_google_callback(self, mock_validate_and_return_user: Mock):
+        response_redirect = urlparse(response.url)
+        self.assertEqual(response_redirect.scheme, expected_redirect_url.scheme, msg)
+        self.assertEqual(response_redirect.netloc, expected_redirect_url.netloc, msg)
+        # path does not matches i tried it manually though it should fail right?
```
what do you mean here? is this comment still valid?

mlodic commented on authentication/views.py:
```diff
@@ -285,16 +285,16 @@ def checkConfiguration(request):
-        raise AuthenticationFailed("GitHub OAuth is not configured.")
-
     try:
+        if client_id is None or client_id == "":
+            raise AuthenticationFailed("GitHub OAuth is not configured.")
+
```
I don't really get why you move code randomly where I asked questions...can we have conversations from my comments?

shivam-Purohit commented:
Also, I am confused with the deepsource python error. I reverted the file to its default state and then added both the path again.
I just press enter from the last line end to get to new line and added the path. Still its the same and I even cant push anything cause for git its the same file. Nothing changed.

mlodic commented:
> Also, I am confused with the deepsource python error. I reverted the file to its default state and then added both the path again.
I just press enter from the last line end to get to new line and added the path. Still its the same and I even cant push anything cause for git its the same file. Nothing changed.

Nevermind, ignore this
